hphp/runtime/vm/jit/vasm-reuseimm.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+   +----------------------------------------------------------------------+
+   | HipHop for PHP                                                       |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2010-present Facebook, Inc. (http://www.facebook.com)  |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+*/
+
+
+#include "hphp/runtime/vm/jit/vasm.h"
+
+#include "hphp/runtime/vm/jit/vasm-gen.h"
+#include "hphp/runtime/vm/jit/vasm-instr.h"
+#include "hphp/runtime/vm/jit/vasm-print.h"
+#include "hphp/runtime/vm/jit/vasm-reg.h"
+#include "hphp/runtime/vm/jit/vasm-unit.h"
+#include "hphp/runtime/vm/jit/vasm-util.h"
+
+#include "hphp/util/arch.h"
+#include "hphp/util/assertions.h"
+
+#include <cstdlib>
+
+TRACE_SET_MOD(vasm);
+
+namespace HPHP { namespace jit {
+
+namespace {
+
+// track ldimmq values
+struct qTrack {
+  Immed64 val;
+  Vreg base;
+  int ttl;
+};
+
+struct Env {
+  Vunit& unit;
+  int tick;
+  std::deque<qTrack> qdeque; // leaky bucket of recent immediates
+};
+
+bool isMultiword(int64_t imm) {
+  switch(arch()) {
+  case Arch::X64:
+  case Arch::PPC64:
+    break;
+  case Arch::ARM:
+    uint64_t val = std::abs(imm);
+    if (val > (1 << 16)) return true;
+    break;
+  }
+  return false;
+}
+
+// candidate around +/- uimm12
+bool reuseCandidate(Env& env, int64_t p, Vreg& reg, int& delta) {
+  delta = 0;
+  for (auto const& itr : env.qdeque) {
+    int64_t q = itr.val.q();
+    if ((p >= q) && (p < (q + 4095))) {
+      delta = safe_cast<int>(p-q);
+      reg = itr.base;
+      return true;
+    }
+    if ((p < q) && (q < (p + 4095))) {
+      delta = -safe_cast<int>(q-p);
```

Comment:
`q - p`

Except also, why can't this just be `delta = safe_cast<int>(p - q)`?  Both of them are signed integers within a certain small delta from one another, so that should be fine, no?